presto-hive/src/main/java/com/facebook/presto/hive/HiveMaterializedViewUtils.java

```diff
@@ -0,0 +1,288 @@
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.lang.String.format;
+import static java.util.Collections.emptyMap;
+
+public class HiveMaterializedViewUtils
+{
+    private static final MaterializedDataPredicates EMPTY_MATERIALIZED_VIEW_DATA_PREDICATES = new MaterializedDataPredicates(ImmutableList.of(), ImmutableList.of());
+
+    private HiveMaterializedViewUtils()
+    {
+    }
+
+    /**
+     * Validate the partition columns of a materialized view to ensure 1) a materialized view is partitioned; and 2) it has at least one partition
+     * directly mapped to all base tables.
+     * <p>
+     * A column is directly mapped to a base table column if it is derived directly or transitively from the base table column,
+     * by only selecting a column or an aliased column without any function or operator applied.
+     * For example, with SELECT column_b AS column_a, column_a is directly mapped to column_b.
+     * With SELECT column_b + column_c AS column_a, column_a is not directly mapped to any column.
+     * <p>
+     * {@code viewToBaseColumnMap} only contains direct column mappings.
+     */
+    public static void validateMaterializedViewPartitionColumns(
+            SemiTransactionalHiveMetastore metastore,
+            MetastoreContext metastoreContext,
+            Table viewTable,
+            ConnectorMaterializedViewDefinition viewDefinition)
+    {
+        SchemaTableName viewName = new SchemaTableName(viewTable.getDatabaseName(), viewTable.getTableName());
+
+        Map<String, Map<SchemaTableName, String>> viewToBaseColumnMap = viewDefinition.getColumnMappingsAsMap();
+        if (viewToBaseColumnMap.isEmpty()) {
+            throw new PrestoException(
+                    NOT_SUPPORTED,
+                    format("Materialized view %s must have at least one column directly defined by a base table column.", viewName));
+        }
+
+        List<Column> viewPartitions = viewTable.getPartitionColumns();
+        if (viewPartitions.isEmpty()) {
+            throw new PrestoException(NOT_SUPPORTED, "Unpartitioned materialized view is not supported.");
+        }
+
+        List<Table> baseTables = viewDefinition.getBaseTables().stream()
+                .map(baseTableName -> metastore.getTable(metastoreContext, baseTableName.getSchemaName(), baseTableName.getTableName())
+                        .orElseThrow(() -> new TableNotFoundException(baseTableName)))
+                .collect(toImmutableList());
+
+        Map<SchemaTableName, List<Column>> baseTablePartitions = baseTables.stream()
+                .collect(toImmutableMap(
+                        table -> new SchemaTableName(table.getDatabaseName(), table.getTableName()),
+                        Table::getPartitionColumns));
+
+        for (SchemaTableName baseTable : baseTablePartitions.keySet()) {
+            if (!isCommonPartitionFound(baseTable, baseTablePartitions.get(baseTable), viewPartitions, viewToBaseColumnMap)) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.